### ISSUE TITLE
Fix possible crashes in libutil

### ISSUE THREAD

**Issue by Flamefire** [Return-To-The-Roots/s25client#258]
We need to implement or forbid copy-operators (constructor, assignment) for Socket::HostAddr to avoid undefined behaviour when it gets moved/copied in e.g. std::vector

**Comment by uqs** [comment]
Uh, can you explain what the problem is with this bug, actually? As I see it (Disclaimer: C++ noob, so please ELI5):

struct HostAddr's ctor runs UpdateAddr which will (now that it skips calling freeaddrinfo on nullptr) call getaddrinfo for everything != localhost (this bit seems bogus). The worst thing in there IMHO is that ipv4/ipv6 is handled with a flag, and the code is not address type agnostic.

The dtor will always call freeaddrinfo for these cases. As long as we match all object construction and destruction, there is no leak and this is fine.

Sure, we might call the DNS resolver too often (N squared), but realistically, that will never happen.

Bigger question: Why is there even an Socket::HostToIp() function? We should just call getaddrinfo during Socket::Connect and not care about storing the resulting IP. Right? (I have no idea how SOCKS works in this regard, but I thought you can just pass through the hostnames to the proxy and it will then do the name resolution for you).

A quick fix for not doing the copy could be to not have the vector as a return type, but pass it along by reference? Or would using a const std::vector<HostAddr> ips in Socket::Connect allow the compiler to optimize the copies away? C++ is confusing ...

**Comment by Flamefire** [comment]
The bug is: Whenever you have a destructor, copy constructor or copy assignment operator you need all of them. The assignment is missing which can lead to bugs that are almost impossible to find.
Note that UpdateAddr is not called in the ctor but in the copy ctor!

However: It seems UpdateAddr is never called but in the copy constructor. and addr seems never be accessed (at least in Socket.cpp) so it might be, that we don't need that member. Others can be copied without any problems.
Maybe we can remove the struct altogether and replace it by something else, but got to check the whole source and could also use some reasoning from @Flow86 on the purpose of the struct.

**Comment by Flow86** [comment]
well, the reason was that the client should try all adresses returned by the hostname

**Comment by Flamefire** [comment]
Then I'd just make it as lean as possible: IP, Port, IsIPv6. Nothing more and all will be fine.

BTW: I've just seen, that Socket has a destroying copy constructor. So a "sockFoo = sockBar" leads to different states of those 2. This should not be done as it can lead to bugs, that are really hard to find. If a copy is not possible, just put unimplemented copy assignment/ctor into private.

**Comment by Flamefire** [comment]
uqs: Check my solution on that: I use HostAddr for storing simple information. When the getaddr stuff is required, I use a class the follows RAII (it manages the resource) This class cannot be copied (it doesn't need to ATM) and provides only reference access to its member. No way a user can free that (without doing some nasty stuff) and the intention is clear.
I'm going to address some serious issues in the next couple of commits, like missing copy ctors/assignment, buffer overflows etc. E.g. BurnedWarehouse has an Out-Of-Bounds access on save/load. Policy: No delete[], boost-array instead of c-array

**Comment by uqs** [comment]
Thanks for the explanation (though I'm not sure I fully understand your fix yet)

**Comment by Flamefire** [comment]
Well in a few words: the result of getaddr is contained in a class, it is gotten in the ctor and free'd in the dtor. The class cannot be copied, so there is exactly 1 owner. This ensures, it is free'd exactly once.

Similar the Socket class: I made it a smart reference: It is ref-counted, so the last one who has a reference to a specific socket closes it. --> Exactly 1 close.
